python/ray/dag/tests/experimental/test_cpu_communicator_dag.py

```diff
@@ -0,0 +1,375 @@
+import os
+import sys
+import torch
+
+import pytest
+
+import ray
+import ray.cluster_utils
+from ray.exceptions import RayChannelError
+from ray.experimental.channel.torch_tensor_type import TorchTensorType
+from ray.experimental.channel.cpu_nccl_group import CPUNcclGroup, start_nccl_mock
+from ray.dag import InputNode
+import ray.experimental.collective as collective
+from ray.dag.output_node import MultiOutputNode
+from ray.tests.conftest import *  # noqa
+
+
+@ray.remote
+class CPUTorchTensorWorker:
+    def start_cuda_and_torch_mock(self):
+        start_nccl_mock()
+
+    def send(self, shape, dtype, value: int, send_tensor=True):
+        if not send_tensor:
+            return 1
+        return torch.ones(shape, dtype=dtype) * value
+
+    def send_dict(self, entries):
+        results = {}
+        for key, entry in entries.items():
+            value, shape, dtype = entry
+            results[key] = torch.ones(shape, dtype=dtype) * value
+        return results
+
+    def send_or_raise(self, shape, dtype, value: int, raise_exception=False):
+        if raise_exception:
+            raise RuntimeError()
+        return torch.ones(shape, dtype=dtype) * value
+
+    def recv(self, tensor):
+        # assert tensor.device == self.device
+        return (tensor[0].item(), tensor.shape, tensor.dtype)
+
+    def recv_dict(self, tensor_dict):
+        vals = {}
+        for i, tensor in tensor_dict.items():
+            vals[i] = self.recv(tensor)
+        return vals
+
+    def compute_with_tuple_args(self, args, i: int):
+        shape, dtype, value = args[i]
+        tensor = torch.ones(shape, dtype=dtype) * value
+        return tensor
+
+    def recv_tensor(self, tensor):
+        return tensor
+
+    def return_tensor(self, size: int) -> torch.Tensor:
+        return torch.ones(size)
+
+
+@pytest.mark.parametrize(
+    "ray_start_cluster",
+    [
+        {
+            "num_cpus": 2,
+            "num_gpus": 0,
+            "num_nodes": 1,
+        }
+    ],
+    indirect=True,
+)
+def test_p2p_basic(ray_start_cluster):
+    sender = CPUTorchTensorWorker.remote()
+    receiver = CPUTorchTensorWorker.remote()
+
+    cpu_group = CPUNcclGroup(2, [sender, receiver])
+    ray.get(
+        [
+            sender.start_cuda_and_torch_mock.remote(),
```

Comment:
Let's update this to not require starting the mock first.

Reply:
fixed